src/seraphis_crypto/sp_crypto_utils.cpp

```diff
@@ -0,0 +1,285 @@
+{
+    while (n--)
+        sc_mul(y.bytes, y.bytes, y.bytes);
+    sc_mul(y.bytes, y.bytes, x.bytes);
+    return y;
+}
+//-------------------------------------------------------------------------------------------------------------------
+//-------------------------------------------------------------------------------------------------------------------
+rct::key minus_one()
+{
+    return MINUS_ONE;
+}
+//-------------------------------------------------------------------------------------------------------------------
+rct::key invert(const rct::key &x)
+{
+    CHECK_AND_ASSERT_THROW_MES(!(x == ZERO), "Cannot invert zero!");
+
+    rct::key _1, _10, _100, _11, _101, _111, _1001, _1011, _1111;
+
+    _1 = x;
+    sc_mul(_10.bytes, _1.bytes, _1.bytes);
+    sc_mul(_100.bytes, _10.bytes, _10.bytes);
+    sc_mul(_11.bytes, _10.bytes, _1.bytes);
+    sc_mul(_101.bytes, _10.bytes, _11.bytes);
+    sc_mul(_111.bytes, _10.bytes, _101.bytes);
+    sc_mul(_1001.bytes, _10.bytes, _111.bytes);
+    sc_mul(_1011.bytes, _10.bytes, _1001.bytes);
+    sc_mul(_1111.bytes, _100.bytes, _1011.bytes);
+
+    rct::key inv;
+    sc_mul(inv.bytes, _1111.bytes, _1.bytes);
+
+    inv = sm(inv, 123 + 3, _101);
+    inv = sm(inv, 2 + 2, _11);
+    inv = sm(inv, 1 + 4, _1111);
+    inv = sm(inv, 1 + 4, _1111);
+    inv = sm(inv, 4, _1001);
+    inv = sm(inv, 2, _11);
+    inv = sm(inv, 1 + 4, _1111);
+    inv = sm(inv, 1 + 3, _101);
+    inv = sm(inv, 3 + 3, _101);
+    inv = sm(inv, 3, _111);
+    inv = sm(inv, 1 + 4, _1111);
+    inv = sm(inv, 2 + 3, _111);
+    inv = sm(inv, 2 + 2, _11);
+    inv = sm(inv, 1 + 4, _1011);
+    inv = sm(inv, 2 + 4, _1011);
+    inv = sm(inv, 6 + 4, _1001);
+    inv = sm(inv, 2 + 2, _11);
+    inv = sm(inv, 3 + 2, _11);
+    inv = sm(inv, 3 + 2, _11);
+    inv = sm(inv, 1 + 4, _1001);
+    inv = sm(inv, 1 + 3, _111);
+    inv = sm(inv, 2 + 4, _1111);
+    inv = sm(inv, 1 + 4, _1011);
+    inv = sm(inv, 3, _101);
+    inv = sm(inv, 2 + 4, _1111);
+    inv = sm(inv, 3, _101);
+    inv = sm(inv, 1 + 2, _11);
+
+    // Confirm inversion
+    rct::key temp;
+    sc_mul(temp.bytes, x.bytes, inv.bytes);
+    CHECK_AND_ASSERT_THROW_MES(temp == ONE, "Scalar inversion failed!");
+
+    return inv;
+}
+//-------------------------------------------------------------------------------------------------------------------
+void decompose(const std::size_t val, const std::size_t base, const std::size_t size, std::vector<std::size_t> &r_out)
+{
+    CHECK_AND_ASSERT_THROW_MES(base > 1, "Bad decomposition parameters!");
+    CHECK_AND_ASSERT_THROW_MES(size > 0, "Bad decomposition parameters!");
+    CHECK_AND_ASSERT_THROW_MES(r_out.size() == size, "Bad decomposition result vector size!");
+
+    std::size_t temp = val;
+
+    for (std::size_t i = 0; i < size; ++i)
+    {
+        std::size_t slot = std::pow(base, size - i - 1);
+        r_out[size - i - 1] = temp/slot;
+        temp -= slot*r_out[size - i - 1];
+    }
```

Comment:
You can remove the `size > 0` constaint on this implementation since no undefined behavior will occur if `size == 0`. If size being strictly positive is a concern for callers, then the callers should assert that. In a similar vein, it depends if the implicit resize is fine to a paranoid caller, since `r_out` is always resized to `size`, the caller can reasonably expect the output to be of size `size`. If you make it part a requirement of the function to actually resize it to that value, then an implicit resize should be fine IMO, but that's a matter of personal preference.


I'm confused by what you mean by "right justity" the digits. My impl should match the same behavior. IIUC, your impl also doesn't right-justify, it left-justifies. If `base^(size-1)` > `val`, the "right hand size" (higher indices, MSB side) will be filled with 0s.

Reply:
Could we drop the `size` parameter then?

Reply:
Ok, updated the suggestion diff